File: BST/J.py
class Node:
    def __init__(self, data):
        self.data = data
        self.count = 1
        self.left = None
        self.right = None


class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, node: Node, x):
        if node is None:
            return Node(x)

        if x < node.data:
            node.left = self.insert(node.left, x)
        else:
            node.right = self.insert(node.right, x)

        return node

    def search(self, node: Node, x):
        if node is None:
            return False

        if x < node.data:
            return self.search(node.left, x)
        elif x > node.data:
            return self.search(node.right, x)
        else:
            node.count += 1

        return node

    def get_max(self, node: Node):
        cur_node = node
        while cur_node.right is not None:
            cur_node = cur_node.right

        return cur_node.data

    def print_sorted_tree(self, node: Node):
        if node is None:
            return

        self.print_sorted_tree(node.left)
        print(node.data, node.count)
        self.print_sorted_tree(node.right)


a = list(map(int, input().split()))
a.pop()

tree = BinaryTree()
for item in a:
    if tree.search(tree.root, item) == False:
        tree.root = tree.insert(tree.root, item)

tree.print_sorted_tree(tree.root)

File: BST/J.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Node
{
    int data;
    int count;
    Node *left;
    Node *right;

    Node(int data) : data(data), count(1), left(nullptr), right(nullptr) {}
};

class BinaryTree
{
public:
    Node *root;

    BinaryTree() : root(nullptr) {}

    Node *insert(Node *node, int x)
    {
        if (node == nullptr)
        {
            return new Node(x);
        }

        if (x < node->data)
        {
            node->left = insert(node->left, x);
        }
        else
        {
            node->right = insert(node->right, x);
        }

        return node;
    }

    Node *search(Node *node, int x)
    {
        if (node == nullptr)
        {
            return nullptr;
        }

        if (x < node->data)
        {
            return search(node->left, x);
        }
        else if (x > node->data)
        {
            return search(node->right, x);
        }
        else
        {
            node->count++;
            return node;
        }
    }

    int get_max(Node *node)
    {
        Node *cur_node = node;
        while (cur_node->right != nullptr)
        {
            cur_node = cur_node->right;
        }
        return cur_node->data;
    }

    void print_sorted_tree(Node *node)
    {
        if (node == nullptr)
        {
            return;
        }

        print_sorted_tree(node->left);
        cout << node->data << " " << node->count << endl;
        print_sorted_tree(node->right);
    }
};

int main()
{
    vector<int> a;
    int input;
    cin >> input;
    while (input != 0)
    {
        a.push_back(input);
        cin >> input;
    }

    BinaryTree tree;
    for (int item : a)
    {
        if (tree.search(tree.root, item) == nullptr)
        {
            tree.root = tree.insert(tree.root, item);
        }
    }

    tree.print_sorted_tree(tree.root);

    return 0;
}